From fdddadfcaa9f2f6034d0d830c12b0c59b8453b02 Mon Sep 17 00:00:00 2001
From: Siye Liu <siliu@microsoft.com>
Date: Tue, 04 Jun 2024 20:08:34 +0000
Subject: [PATCH] Prevent crash in |InputMethodController::SetComposition|

Layout can be dirty again after calling |SelectComposition| in
|InputMethodController::SelectComposition|. There is existing code to
update layout in the method later. We should move update style call
right after |SelectComposition|.

Bug: 336992377
Change-Id: I7bdf207e4ba25d76519b02d8749b27f193bc06d0
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5594478
Commit-Queue: Siye Liu <siliu@microsoft.com>
Reviewed-by: Anupam Snigdha <snianu@microsoft.com>
Cr-Commit-Position: refs/heads/main@{#1310155}
---

--- a/third_party/blink/renderer/core/editing/ime/input_method_controller.cc
+++ b/third_party/blink/renderer/core/editing/ime/input_method_controller.cc
@@ -979,6 +979,10 @@
 
   SelectComposition();
 
+  // TODO(editing-dev): The use of UpdateStyleAndLayout
+  // needs to be audited. see http://crbug.com/590369 for more details.
+  GetDocument().UpdateStyleAndLayout(DocumentUpdateReason::kEditing);
+
   if (GetFrame().Selection().ComputeVisibleSelectionInDOMTree().IsNone()) {
     return;
   }
@@ -987,10 +991,6 @@
   if (!target)
     return;
 
-  // TODO(editing-dev): The use of UpdateStyleAndLayout
-  // needs to be audited. see http://crbug.com/590369 for more details.
-  GetDocument().UpdateStyleAndLayout(DocumentUpdateReason::kEditing);
-
   PlainTextRange selected_range = CreateSelectionRangeForSetComposition(
       selection_start, selection_end, text.length());
 
